sympy/simplify/cse_main.py

```diff
@@ -536,3 +536,101 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
         return replacements, reduced_exprs
 
     return postprocess(replacements, reduced_exprs)
+
+
+class Marker:
+    pass
+
+# XXX: Should this go in sympy.utilities.iterables?
+def shortest_repeated_subsequence(S, ignore=(Marker,)):
```

Comment:
Actually, in my tests, a potential problem with this algorithm is that it requires many iterations. When I tested against random strings, it often generated tons of replacements. But it does seem to result in fewer operations (when the sequences are treated as products) than when using the maximum subsequences. 

Here is an example (lowercase means the original string and uppercase means a common substring)

```
>>> strs = [
 'jkbbnlglij',
 'hicicfhdkk',
 'hbobhgbipa',
 'epiifchlck',
 'afglcdmica',
 'dmdakhkmdk',
 'gajlodnpmc',
 'jpodjhkopc',
 'kmkoaacchb',
 'jlghgkcobg']
>>> s, r = cse_strs(strs, 'min3')
Starting ops 90
Ending ops 67 + 11 = 78
>>> s, r
(['jkbbnClij',
  'hJJfhHk',
  'DABbipa',
  'epiifchKk',
  'afgKIJa',
  'IdakhFH',
  'gajlGnpmc',
  'jpGjhEpc',
  'FEaaccD',
  'jCBkcAg'],
 {'A': 'ob',
  'B': 'hg',
  'C': 'lg',
  'D': 'hb',
  'E': 'ko',
  'F': 'km',
  'G': 'od',
  'H': 'dk',
  'I': 'dm',
  'J': 'ic',
  'K': 'lc'})
```